arclines/holy/patterns.py

```diff
@@ -222,3 +223,182 @@ def score_quad_matches(fidx):
+                        lup = sz_l
+                    for ll in range(l+nptn-1, lup):
+                        for xl in range(l+1, ll):
+                            lval = (linelist[xl]-linelist[l])/(linelist[ll]-linelist[l])
+                            tst = lval-dval
+                            if tst < 0.0:
+                                tst *= -1.0
+                            if tst <= tol:
+                                lindex[cntdet, 0] = l
+                                lindex[cntdet, 1] = xl
+                                lindex[cntdet, 2] = ll
+                                dindex[cntdet, 0] = d
+                                dindex[cntdet, 1] = xd
+                                dindex[cntdet, 2] = dd
+                                tst = (linelist[ll]-linelist[l]) / (detlines[dd]-detlines[d])
+                                wvcen[cntdet] = (npixels/2.0) * tst + (linelist[ll]-tst*detlines[dd])
+                                disps[cntdet] = tst
+                            cntdet += 1
+    return dindex, lindex, wvcen, disps
+
+
+def solve_triangles(detlines, linelist, dindex, lindex, best_dict=None):
+    """  Given a starting solution, find the best match for all detlines
+
+    Parameters
+    ----------
+    detlines : ndarray
+      list of detected lines in pixels (sorted, increasing)
+    linelist : ndarray
+      list of lines that should be detected (sorted, increasing)
+    dindex : ndarray
+      Index array of all detlines used in each triangle
+    lindex : ndarray
+      Index array of the assigned line to each index in dindex
+    best_dict : dict
+      Contains all relevant details of the fit
+
+    Returns
+    -------
+
+    """
+    nlines = detlines.size
+    if best_dict is None:
+        best_dict = dict(nmatch=0, ibest=-1, bwv=0.)
+
+    # Find the best ID of each line
+    detids = np.zeros(nlines)
+    scores = ['None' for xx in range(nlines)]
+    mask = np.zeros(nlines, dtype=np.bool)
+    ngd_match = 0
+    for dd in range(nlines):
+        ww = np.where(dindex == dd)
+        if ww[0].size == 0:
+            continue
+        unq, cnts = np.unique(lindex[ww], return_counts=True)
+        detids[dd] = linelist[unq[np.argmax(cnts)]]
+        scr = score_triangles(cnts)
+        scores[dd] = scr
+        if scr in ["Perfect", "Very Good", "Good", "OK"]:
+            mask[dd] = True
+            ngd_match += 1
+
+    # Iteratively fit this solution, and ID all lines.
+    if ngd_match > best_dict['nmatch']:
+        best_dict['mask'] = mask
+        best_dict['nmatch'] = ngd_match
+        best_dict['scores'] = scores
+        best_dict['IDs'] = detids
+    return
+
+
+def score_triangles(counts):
+    """  Grades for the triangle results
+    Parameters
+    ----------
+    fidx
+
+    Returns
+    -------
+    scores : list
+
+    """
+    ncnt = counts.size
+    max_counts = np.max(counts)
+    sum_counts = np.sum(counts)
+    # Score
+    if (ncnt == 1) & (max_counts >= 4):
```

Comment:
Maybe we should set these values up in the *def*.

Reply:
I thought about putting this in a dictionary (at least the cases except `'Perfect'`), but figured this was more readable. However, if you can think of a better alternative, I'm all for it!

Reply:
Nah, I agree.  Looks more readable.